### PR TITLE
Encode all join conditions in a single expression field

### LINKED ISSUE #7611: Substrait: Combine join on and filter expressions in a single Substrait JoinRel's field
### Is your feature request related to a problem or challenge?

As of version 31.0.0, `datafusion` has a join [on](https://github.com/apache/arrow-datafusion/blob/main/datafusion/expr/src/logical_plan/plan.rs#L2214C11-L2214C11) field which is used for equi-join conditions and a join [filter](https://github.com/apache/arrow-datafusion/blob/main/datafusion/expr/src/logical_plan/plan.rs#L2216) field for non-equi-join conditions.

Currently, the producer puts the non-equi-join conditions in the Substrait [post_join_filter](https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L161). However, we can also combine all conditions and put it in the [expression](https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L160) field of the JoinRel in Substrait.

Please refer to the motivation in the following comment. ~The motivation behind this change request is that this will let other DB systems decide what to do with the entire condition, as opposed to having to process them separately. Right now, if there is no equal condition, the producer will output just `Literal(True)` as the join expression, and put the rest of the condition in the `post_join_filter`. Having a redundant `Literal` expression adds unnecessary overhead of evaluating this condition. It also implies that you’re performing a cartesian product THEN a filter, as opposed to just a non-equi-join, which does not completely align with the original plan intent.~

### Describe the solution you'd like

All valid join conditions to be encoded in the Substrait `JoinRel`'s [expression](https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L160) field.

### Describe alternatives you've considered

None.
~The current approach works correctly semantically, but it can make downstream query execution inefficient.~

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nseekhao**
## Which issue does this PR close?

Closes #7611 .

## Rationale for this change

For ~more efficient~ fixing the incorrect representation of join condition. Please refer to the related issue for more details.

## What changes are included in this PR?

- Producer: instead of encoding the join `on` and `filter` separately in `expression` and `post_join_filter`, this PR encodes both in the `expression` field of the `JoinRel`.
- Consumer: Adds a check for in each part of the conjunctive JoinRel expression. If the sub-expression is a `Column Eq Column` expression, then it gets added to the `on` field, otherwise, it gets AND-ed with the `filter` field.

## Are these changes tested?

Yes.

No new tests were added. Since the producer now hardcodes `post_join_filter` as `None`, if the changes are not working correctly, the existing would fail.

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (Dandandan) [inline datafusion/substrait/src/logical_plan/consumer.rs:372]: I'm not sure if this is a good idea, it seems to make things more complex.

Comment (nseekhao) [reply datafusion/substrait/src/logical_plan/consumer.rs:372]: @Dandandan Thank you for pointing this out. I agree that this makes things more complicated, and my apologies for the inaccurate explanation of the issue. I added a comment to correct my description of the [issue](https://github.com/apache/arrow-datafusion/issues/7611#issuecomment-1729906548).

The high-level idea is that join `filter` and `post_join_filter` do not have the same meaning semantically. The former is for filtering input during/pre join, the latter is for filtering the output of the join (post-join). Currently in `datafusion`, we do not have a field in join relation that represents a post-join predicate (the parser/logical optimizer takes care of creating an appropriate filter relation if necessary). So the producer should only generate plans with `None` as `post_join_filer`.

I'll modify the consumer to throw an error for now if there's a `post_join_filter`. Later, we can wrap the join relation with a filter relation if we want to support `post_join_filter`, or if the later version of `datafusion` supports `post_join_filter` directly in the join relation, then we can also add the support in both producer and consumer as well.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (alamb) [inline datafusion/substrait/src/logical_plan/consumer.rs:345]: 👍

File: datafusion/substrait/src/logical_plan/producer.rs
Comment (alamb) [inline datafusion/substrait/src/logical_plan/producer.rs:306]: Perhaps you could use `conjunction` here and simplify the code https://docs.rs/datafusion/latest/datafusion/optimizer/utils/fn.conjunction.html

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (alamb) [reply datafusion/substrait/src/logical_plan/consumer.rs:370]: I looked into the substrait spec, and it doesn't really talk about what semantics of the `post_join_filter` is: https://substrait.io/relations/logical_relations/#join-operation
https://github.com/search?q=repo%3Asubstrait-io%2Fsubstrait%20post_join_filter&type=code

There is a subtle distinction between non equality filters applied *during* the join (in the `ON` clause) and applied post join for non-INNER joins: for non inner joins the filters during the join don't filter out input rows (they still come out, just with NULL matches)

> So the producer should only generate plans with None as post_join_filer. 

This makes sense to me

For the consumer, there is already code in DataFusion that breaks up an arbitrary `Expr` into equality predicates and others. This is how the SQL frontend creates a Join (a single expr):

https://github.com/apache/arrow-datafusion/blob/a514b6752b063a5a3006aa114297520a933339b0/datafusion/sql/src/relation/join.rs#L134-L141

I think we could do the same here in the subtrait consumer which would be much simpler, and would let the normal DataFusion optimization machinery work.

**Review by alamb — COMMENTED** [review]
Thank you @nseekhao and @Dandandan 

I think this code could be made simpler and left  some suggestions about how to do so. 

That being said the only thing I think is needed prior to merge is some tests (so we don't break this code in some future refactoring). I don't think the simplifications are necessary

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (nseekhao) [reply datafusion/substrait/src/logical_plan/consumer.rs:370]: Thanks @alamb for the pointer! I'll look into this and update the consumer.

**Comment by nseekhao** [comment]
> Thank you @nseekhao and @Dandandan
> 
> I think this code could be made simpler and left some suggestions about how to do so.
> 
> That being said the only thing I think is needed prior to merge is some tests (so we don't break this code in some future refactoring). I don't think the simplifications are necessary

I don't mind simplifying the code so I'll work on it following your earlier suggestions.

As for the tests, I agree. Since, existing cases already test for mixed join condition ([roundtrip_non_equi_join()](https://github.com/apache/arrow-datafusion/blob/main/datafusion/substrait/tests/cases/roundtrip_logical_plan.rs#L387)) and non-equi-only join ([roundtrip_non_equi_inner_join()](https://github.com/apache/arrow-datafusion/blob/main/datafusion/substrait/tests/cases/roundtrip_logical_plan.rs#L381)), are you thinking a new set of tests that specifically test for the absence of `post_join_filter`? Or did you have something else in mind? Thanks!

**Comment by alamb** [comment]
> I don't mind simplifying the code so I'll work on it following your earlier suggestions.

Thank you! BTW I found the relevant code to extract equijoins predicates here: https://docs.rs/datafusion/latest/datafusion/optimizer/extract_equijoin_predicate/struct.ExtractEquijoinPredicate.html

> are you thinking a new set of tests that specifically test for the absence of post_join_filter

Yes I guess I was thinking about that. Mostly I was thinking "if we broke / reverted the code in this PR accidentally, how would we know something wasn't correct"? Without some test coverage the behavior could change and we don't know.

**Comment by alamb** [comment]
Marking as draft as I believe this PR is not waiting for any feedback (there are some comments to address). Thanks again for pushing this forward @nseekhao

**Comment by nseekhao** [comment]
@alamb I extracted the predicate splits from the `from_substrait_rel()` so the code is easier to read. Unfortunately, I couldn't use the function `split_eq_and_noneq_join_predicate()` (from your [ref](https://docs.rs/datafusion/latest/datafusion/optimizer/extract_equijoin_predicate/struct.ExtractEquijoinPredicate.html)) directly because the function will put the equal condition with `nulls_equal_nulls = true` (`IsNotDistinctFrom`) into a join filter, as opposed to putting it in the join condition.

As for the test, I added a test function to be used when join has a non-equi condition, to ensure that the join filter does not mistakenly get embedded into a Substrait plan as a `post_join_filter`. LMK if this is what you had in mind, otherwise we can discuss further changes. Thanks for the review!

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (alamb) [inline datafusion/substrait/src/logical_plan/consumer.rs:403]: Given that the [`ExtractEquijoinPredicate`](https://github.com/apache/arrow-datafusion/blob/093b775adc3593e9e5cb7343e28406ed458551ad/datafusion/optimizer/src/extract_equijoin_predicate.rs#L31-L41) optimizer pass already  splits up join predicates into equijoin predicates and "other" predicates, I wonder if simply create the `LogicalPlan::Join` using `join.expression` (and let the subsequent optimizer pass sort it out)? 

Something like

```rust
left.join(
  right.build()?,
  join_type,
  (vec![], vec![]),
  on, // <-- use the filter directly here, let optimizer pass extract the equijoin columns
  nulls_equal_nulls,
)?
```

It makes me realize when looking at the API for [`LogicalPlanBuilder::join`](https://docs.rs/datafusion/latest/datafusion/logical_expr/struct.LogicalPlanBuilder.html#method.join) that the API is super confusing. It would be nice to improve that API to make it clear that a `join` can just take a single `Expr` and DataFusion will sort out figuring out the join columns, etc.

File: datafusion/substrait/tests/cases/roundtrip_logical_plan.rs
Comment (alamb) [inline datafusion/substrait/tests/cases/roundtrip_logical_plan.rs:626]: Looks like it might be helpful (eventually) do define `TreeNode` for `Rel` to implement walking efficiently.

**Review by alamb — APPROVED** [review]
Thank you @nseekhao  -- I think we could merge this PR as is. I still think it could perhaps be made even simpler (by avoiding the join predicate analysis). Let me know what you think.

File: datafusion/substrait/src/logical_plan/consumer.rs
Comment (alamb) [reply datafusion/substrait/src/logical_plan/consumer.rs:403]: It turns out this is exactly what `DataFrame::join_on` does -- I have filed a ticket with a way to make this clearer: https://github.com/apache/arrow-datafusion/issues/7766#issue-1931312146

**Review by alamb — APPROVED** [review]
I think this PR is better that what is on master, so I plan to merge it in once CI passes. I will propose a follow on PR that uses the newly added `LogicalPlanBuilder::join_on` to simplify it subsequently

**Comment by alamb** [comment]
I have merged up from main and plan to merge this PR once CI passes

**Comment by alamb** [comment]
>  I will propose a follow on PR that uses the newly added LogicalPlanBuilder::join_on to simplify it subsequently

I tried this and it turns out there is some subtlety involved with `a = a` vs `a IS DISTINCT FROM` that I was not able to solve when trying to simplify.

**Comment by alamb** [comment]
Thanks again @nseekhao

**Pull request merged**
